include/VectorGraph.h

```diff
@@ -0,0 +1,739 @@
+	void styleChanged();
+public slots:
+	void dataArrayChanged();
+	void updateGraphModel(bool shouldUseGetLastSamples);
+	void dataArrayClearedEvent(int arrayId);
+	void dataArrayStyleChanged();
+private:
+	std::vector<VectorGraphDataArray> m_dataArrays;
+	unsigned int m_maxLength;
+
+	// block threads that want to access
+	// a dataArray's getSamples() at the same time
+	QMutex m_getSamplesAccess;
+	QMutex m_bakedSamplesAccess;
+};
+
+class LMMS_EXPORT VectorGraphDataArray
+{
+
+public:
+	// avoid using this or run updateConnections() after initialization
+	VectorGraphDataArray();
+	VectorGraphDataArray(
+	bool isFixedSize, bool isFixedX, bool isFixedY, bool isNonNegative,
+	bool isFixedEndPoints, bool isSelectable, bool isEditableAttrib, bool isAutomatableEffectable,
+	bool isSaveable, VectorGraphModel* parentIn, int arrayId);
+	~VectorGraphDataArray();
+
+	void updateConnections(VectorGraphModel* parentIn);
+
+	// see descriptions in privete
+	void setIsFixedSize(bool bValue);
+	void setIsFixedX(bool bValue);
+	void setIsFixedY(bool bValue);
+	void setIsFixedEndPoints(bool bValue);
+	void setIsSelectable(bool bValue);
+	void setIsEditableAttrib(bool bValue);
+	void setIsAutomatableEffectable(bool bValue);
+	void setIsSaveable(bool bValue);
+	void setIsNonNegative(bool bValue);
+	void setLineColor(QColor color);
+	void setActiveColor(QColor color);
+	void setFillColor(QColor color);
+	void setAutomatedColor(QColor color);
+
+	// returns true if successful
+	// if callDataChanged then it will call dataChanged() --> paintEvent()
+	bool setEffectorArrayLocation(int arrayLocation, bool callDataChanged);
+
+	bool getIsFixedSize();
+	bool getIsFixedX();
+	bool getIsFixedY();
+	bool getIsFixedEndPoints();
+	bool getIsSelectable();
+	bool getIsEditableAttrib();
+	bool getIsAutomatableEffectable();
+	bool getIsSaveable();
+	bool getIsNonNegative();
+	QColor* getLineColor();
+	QColor* getActiveColor();
+	QColor* getFillColor();
+	QColor* getAutomatedColor();
+	// returns -1 if it has no effector
+
+	int getEffectorArrayLocation();
+	int getId();
+
+
+	// array: -------------------
+	// returns the location of added/found point, -1 if not found or can not be added
+	int add(float newX);
+	// deletes the point in pointLocation location if m_isFixedSize is disabled
+	void del(unsigned int pointLocation);
+	// clears m_dataArray without any checks
+	inline void clear()
+	{
+		m_dataArray.clear();
+		m_needsUpdating.clear();
+		// m_automationModelArray sould not be cleared without destruction
+		clearedEvent();
+		getUpdatingFromPoint(-1);
+		dataChanged();
+	}
+	inline size_t size()
+	{
+		return m_dataArray.size();
+	}
+	// clamps down the values to 0 - 1, -1 - 1
+	// sorts array, removes duplicated positions, calls dataChanged() if callDataChanged
+	// clamp: should clamp, sort: should sort
+	void formatArray(std::vector<std::pair<float, float>>* dataArrayOut, bool shouldClamp, bool shouldRescale, bool shouldSort, bool callDataChanged);
```

Comment:
After seeing a lot of `std::pair<float, float>` I'm starting to think we should probably add a `using PointF = std::pair<float, float>` for simplicity.